### PR TITLE
 Federated Environment Startup

### PR DESCRIPTION
**Pull request by Baunsgaard**
- Upgrade the startup of the Federated Environment
- Support for Default Port
- Relative and static file path URL for Federated Worker
- Minor Startup cleanup in Federated Worker
- No need for extra file argument to start a federated worker

### CODE REVIEW

**Comment by kev-inn** [comment]
Note: my matrix-matrix-multiplication implementation is ready for a PR, but I already included the commits of this PR, so I will wait until this is merged, include the changes and then do my own PR.

File: src/main/java/org/tugraz/sysds/api/DMLScript.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/api/DMLScript.java:160]: Is there a good reason to remove the try/catch ? If not, please revert that change

File: pom.xml
Comment (corepointer) [inline pom.xml:929]: Is this standalone profile needed by the other changes in this PR? If not, please single it out to another commit/PR to better keep track of unrelated changes.

File: src/main/java/org/tugraz/sysds/conf/DMLConfig.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/conf/DMLConfig.java:92]: why did you change to an already used port?

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorker.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorker.java]: no exception thrown in that method

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java]: You removed the square brackets before (in FederatedWorker.java). If you fancy consistency... ;-)

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java:83]: I know it's not your line of code, but that DMLRuntimeException looks unfitting.

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (corepointer) [inline src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java:287]: Ok I give up. It's a DMLRuntimeException 😜

**Review by corepointer — CHANGES_REQUESTED** [review]
A few minor things (see comments).

File: src/main/java/org/tugraz/sysds/api/DMLScript.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/api/DMLScript.java:160]: The Point here was that this Exception was already handled just in the next function, executeScript() and then thrown again just wrapped in another exception. That is why i moved the DMLScriptException. (look at line 239)
The other exception have exactly the same action, therefore it was moved as well. Since we are using Java Language level 8 we are able to catch multiple specific exceptions in a single catch block.

All that said, i can see i did not finish the work properly, since i am still throwing the exception, in the next method. (which is not needed anymore for same execution)

Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/api/DMLScript.java:160]: Looking further into it, it seems like this part of the code could be cleaned up further in the exception handling, since there is many inconsistencies.

File: src/main/java/org/tugraz/sysds/conf/DMLConfig.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/conf/DMLConfig.java:92]: Since the port i originally chose, was not open by default in default cluster setup, i changed it to one that is known to be open in most cases for a cluster setup.

Other options are available, Link: [OpenPorts](https://www.cloudera.com/tutorials/hortonworks-sandbox-guide/3.html)
But while trying one of the "Port for custom use" (specificly 111) It did not work.

File: pom.xml
Comment (Baunsgaard) [reply pom.xml:929]: This is related to this PR, for the standalone execution of a Federated worker. Since we want to enable the execution on locations that does not have all the dependent packages installed.
It is specifically a optional Profile to compile, since it should not be executed/used by default as of yet, due to the overhead in size of the constructed standalone Jar file.

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorker.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorker.java]: ok

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java]: yes, i did change it ... sometimes, looked through and changed appropriately

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java:83]: Sorry, but it is most fitting here.

File: src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java
Comment (Baunsgaard) [reply src/main/java/org/tugraz/sysds/runtime/controlprogram/federated/FederatedWorkerHandler.java:287]: Yes

**Review by Baunsgaard — COMMENTED** [review]
Reviewed Comments

**Comment by mboehm7** [comment]
LGTM - thanks for the patch @Baunsgaard. As discussed offline, it would be great to have another look into the federated tests.

**Pull request closed without merging**
